File: Number of Subsequences That Satisfy the Given Sum Condition.java
class Solution {
    public int numSubseq(int[] nums, int target) {
        Arrays.sort(nums); // Sort the array in non-decreasing order
        int n = nums.length;
        int mod = 1000000007;
        int left = 0, right = n - 1;
        long ans = 0;
        int[] pow2 = new int[n];
        pow2[0] = 1;
        for (int i = 1; i < n; i++) {
            pow2[i] = (pow2[i - 1] * 2) % mod; // Precompute powers of 2
        }
        while (left <= right) { // Two pointer approach to count subsequences
            if (nums[left] + nums[right] > target) {
                right--; // If sum of min and max is greater than target, decrement right
            } else {
                ans = (ans + pow2[right - left]) % mod; // Compute number of subsequences using precomputed powers of 2
                left++; // Increment left
            }
        }
        return (int) ans; // Return answer modulo 10^9+7
    }
}

File: Number of Subsequences That Satisfy the Given Sum Condition.cpp
class Solution {
public:
    int numSubseq(vector<int>& nums, int target) {
        int n = nums.size();
        int mod = 1e9+7;

        sort(nums.begin(),nums.end());
        int ans = 0,l=0,r=n-1;

        vector<int> power(n, 1);
        for (int i = 1; i < n; i++) {
            power[i] = (power[i - 1] * 2) % mod;
        }

        while(l<=r){
            if(nums[l]+nums[r]<=target){
                ans = (ans + power[r-l])%mod;
                l++;
            }
            else{
                r--;
            }
        }
        return ans;
    }
};
